.github/workflows/test.yaml

```diff
@@ -84,5 +94,48 @@ jobs:
             kubectl rollout status daemonset.apps/image-prefetch --timeout=90s ||
             kubectl describe daemonset.apps/image-prefetch
           }
-      - env: { PGO_TEST_TIMEOUT_SCALE: 1.2 }
-        run: make createnamespaces check-envtest-existing
+      - run: make createnamespaces check-envtest-existing
+        env:
+          PGO_TEST_TIMEOUT_SCALE: 1.2
+          GO_TEST: go test --coverprofile 'envtest-existing.coverage' --coverpkg ./internal/...
+
+      # Upload coverage to GitHub
+      - run: gzip envtest-existing.coverage
+      - uses: actions/upload-artifact@v2
+        with:
+          name: "kubernetes-k3d=${{ matrix.kubernetes }}"
+          path: envtest-existing.coverage.gz
+          retention-days: 1
+
+  coverage-report:
+    runs-on: ubuntu-latest
+    needs:
+      - kubernetes-api
+      - kubernetes-k3d
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2
+        with: { go-version: 1.x }
+      - uses: actions/download-artifact@v2
+        with: { path: download }
+
+      # Combine the coverage profiles by taking the mode line from any one file
+      # and the data from all files. Print a list of functions with less than
+      # 100% coverage, and upload a complete HTML report.
+      - name: Generate report
+        run: |
+          gunzip --keep download/*/*.gz
+          ( sed -e '1q' download/*/*.coverage
+            tail -qn +2 download/*/*.coverage ) > total.coverage
+          go tool cover --func total.coverage -o total-coverage.txt
+          go tool cover --html total.coverage -o total-coverage.html
+          sed -e '/100.0%/d' -e "s,$(go list -m)/,," total-coverage.txt
+          echo "::notice title=Total Coverage::$(awk 'END { print $3 }' total-coverage.txt)"
+
+      # Upload coverage to GitHub
+      - run: gzip total-coverage.html
+      - uses: actions/upload-artifact@v2
+        with:
+          name: coverage-report
+          path: total-coverage.html.gz
+          retention-days: 15
```

Comment:
I chose to upload the HTML report as an artifact because it shows which lines lack coverage. This can go away if the plaintext, function-level report is enough. The default retention of 90 days seems too long, so I chose something shorter here.

Reply:
Is there any limit to how much we can store or for how long?

Reply:
The docs say artifacts "use storage space" and that Actions are "free for public repositories". The docs about billing for Actions talk about what's free for private repositories.

- https://docs.github.com/en/actions/advanced-guides/storing-workflow-data-as-artifacts
- https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions

So, I don't see any hard limits. Perhaps charges will accrue at "$0.25 USD per GB of storage per month."

Reply:
Sorry, the docs _are_ clear about retention. Up to 90 days for public repos.